File: src/core/orchestrator.ts
import { generateRunId, generateSourceId } from './ids.js'
import { detectSource } from './source-detector.js'
import { loadFile } from './file-loader.js'
import { loadFolder } from './folder-loader.js'
import { loadUrl } from './url-loader.js'
import { loadGitHub } from './github-loader.js'
import { chunkTextString } from './chunker.js'
import { extract } from './extractor.js'
import { saveRun, loadRun, listRuns } from './artifact-store.js'
import { syncToMemoryLane } from './memorylane-sync.js'
import { exportRun } from './export-import.js'
import { search } from './search.js'
import { buildMemoryRecords } from './memory-records-builder.js'
import type { ContextLaneSource, ContextLaneRun, ContextLaneChunk } from './schema.js'

export interface IngestOptions {
  input: string
  type?: 'file' | 'folder' | 'url' | 'github' | 'text'
  syncToMemoryLane?: boolean
}

export async function ingest(options: IngestOptions): Promise<ContextLaneRun> {
  const { input, syncToMemoryLane: doSync } = options
  let detectedType = options.type || detectSource(input).type

  const runId = generateRunId()
  const sourceId = generateSourceId()
  const allChunks: ContextLaneChunk[] = []
  const allCitations: ContextLaneCitation[] = []

  const sources: ContextLaneSource[] = []

  if (detectedType === 'file') {
    const file = loadFile(input)
    const label = input.split('/').pop() || input
    sources.push({
      id: sourceId,
      type: 'file',
      input,
      title: label,
      path: input,
      loadedAt: new Date().toISOString(),
    })
    const result = chunkTextString(file.text, sourceId, label, input)
    allChunks.push(...result.chunks)
    allCitations.push(...result.citations)
  }

  if (detectedType === 'folder') {
    const items = loadFolder(input)
    for (const item of items) {
      const itemSourceId = generateSourceId()
      sources.push({
        id: itemSourceId,
        type: 'file',
        input: item.path,
        title: item.relativePath,
        path: item.path,
        loadedAt: new Date().toISOString(),
      })
      const result = chunkTextString(item.text, itemSourceId, item.relativePath, item.path)
      allChunks.push(...result.chunks)
      allCitations.push(...result.citations)
    }
  }

  if (detectedType === 'url') {
    const loaded = await loadUrl(input)
    sources.push({
      id: sourceId,
      type: 'url',
      input,
      title: loaded.title,
      url: input,
      loadedAt: new Date().toISOString(),
    })
    const result = chunkTextString(loaded.text, sourceId, loaded.title || input, undefined, input)
    allChunks.push(...result.chunks)
    allCitations.push(...result.citations)
  }

  if (detectedType === 'github') {
    const loaded = loadGitHub(input)
    for (const item of loaded.items) {
      const itemSourceId = generateSourceId()
      sources.push({
        id: itemSourceId,
        type: 'file',
        input: item.path,
        title: item.relativePath,
        path: item.path,
        loadedAt: new Date().toISOString(),
      })
      const result = chunkTextString(item.text, itemSourceId, item.relativePath, item.path)
      allChunks.push(...result.chunks)
      allCitations.push(...result.citations)
    }
  }

  if (detectedType === 'text') {
    sources.push({
      id: sourceId,
      type: 'text',
      input,
      title: 'inline-text',
      loadedAt: new Date().toISOString(),
    })
    const result = chunkTextString(input, sourceId, 'inline-text')
    allChunks.push(...result.chunks)
    allCitations.push(...result.citations)
  }

  const extraction = extract({ runId, chunks: allChunks, citations: allCitations })
  const memoryRecords = buildMemoryRecords(runId, sources, allChunks, extraction)

  const run: ContextLaneRun = {
    id: runId,
    sources,
    chunks: allChunks,
    extraction: { ...extraction, citations: allCitations },
    memoryRecords,
    createdAt: new Date().toISOString(),
  }

  saveRun(run)

  if (doSync) {
    const syncResult = await syncToMemoryLane(memoryRecords)
    console.error(`MemoryLane sync: ${syncResult.saved} saved, ${syncResult.failed} failed (via ${syncResult.method})`)
  }

  return run
}

export { loadRun, listRuns, search, exportRun, syncToMemoryLane, loadRun as getRun }
export type { ContextLaneRun }

File: python/contextlane/core/orchestrator.py
import os
import time

from .ids import generate_run_id, generate_source_id
from .source_detector import detect_source
from .file_loader import load_file
from .folder_loader import load_folder
from .url_loader import load_url
from .github_loader import load_github
from .chunker import chunk_text_string
from .extractor import extract
from .artifact_store import save_run, load_run, list_runs as _list_runs
from .memorylane_sync import sync_to_memory_lane as _sync
from .export_import import export_run as _export, import_run as _import
from .search import search as _search
from .memory_records_builder import build_memory_records


async def ingest(input_str: str, source_type: str = None, sync_to_memory_lane: bool = False) -> dict:
    if source_type is None:
        source_type, resolved = detect_source(input_str)
    else:
        resolved = input_str

    run_id = generate_run_id()
    all_chunks = []
    all_citations = []
    sources = []

    if source_type == "file":
        file = load_file(resolved)
        label = os.path.basename(resolved)
        source_id = generate_source_id()
        sources.append({"id": source_id, "type": "file", "input": resolved, "title": label, "path": resolved, "loaded_at": time.ctime()})
        result = chunk_text_string(file["text"], source_id, label, resolved)
        all_chunks.extend(result["chunks"])
        all_citations.extend(result["citations"])

    elif source_type == "folder":
        items = load_folder(resolved)
        for item in items:
            item_source_id = generate_source_id()
            sources.append({"id": item_source_id, "type": "file", "input": item["path"], "title": item["relative_path"], "path": item["path"], "loaded_at": time.ctime()})
            result = chunk_text_string(item["text"], item_source_id, item["relative_path"], item["path"])
            all_chunks.extend(result["chunks"])
            all_citations.extend(result["citations"])

    elif source_type == "url":
        loaded = load_url(resolved)
        source_id = generate_source_id()
        sources.append({"id": source_id, "type": "url", "input": resolved, "title": loaded.get("title"), "url": resolved, "loaded_at": time.ctime()})
        result = chunk_text_string(loaded["text"], source_id, loaded.get("title") or resolved, None, resolved)
        all_chunks.extend(result["chunks"])
        all_citations.extend(result["citations"])

    elif source_type == "github":
        loaded = load_github(resolved)
        for item in loaded["items"]:
            item_source_id = generate_source_id()
            sources.append({"id": item_source_id, "type": "file", "input": item["path"], "title": item["relative_path"], "path": item["path"], "loaded_at": time.ctime()})
            result = chunk_text_string(item["text"], item_source_id, item["relative_path"], item["path"])
            all_chunks.extend(result["chunks"])
            all_citations.extend(result["citations"])

    elif source_type == "text":
        source_id = generate_source_id()
        sources.append({"id": source_id, "type": "text", "input": resolved, "title": "inline-text", "loaded_at": time.ctime()})
        result = chunk_text_string(resolved, source_id, "inline-text")
        all_chunks.extend(result["chunks"])
        all_citations.extend(result["citations"])

    extraction = extract(run_id, all_chunks, all_citations)
    memory_records = build_memory_records(run_id, sources, all_chunks, extraction)

    run = {
        "id": run_id,
        "sources": sources,
        "chunks": all_chunks,
        "extraction": extraction,
        "memory_records": memory_records,
        "created_at": time.ctime(),
    }

    save_run(run)

    if sync_to_memory_lane:
        sync_result = await _sync(memory_records)
        print(f"MemoryLane sync: {sync_result['saved']} saved, {sync_result['failed']} failed (via {sync_result['method']})", file=__import__("sys").stderr)

    return run


def load_run_(run_id: str) -> dict:
    return load_run(run_id)


def list_runs() -> list:
    return _list_runs()


def search_(query: str, limit: int = 5) -> list:
    return _search(query, limit)


def export_run_(run_id: str, out_path: str) -> None:
    return _export(run_id, out_path)


def import_run_(in_path: str) -> dict:
    return _import(in_path)


async def sync_memory_lane(run_id: str) -> dict:
    run = load_run(run_id)
    return await _sync(run["memory_records"])
